### ISSUE TITLE
Best way to run external http requests concurrently

### ISSUE THREAD

**Issue by maitham** [tiangolo/fastapi#1280]
I'm currently using the httpx library, and need to make external http requests concurrently. Let's say for example I have a list of urls to map over and need to do this via async await syntax. Whats the best way of adding these requests as tasks to the event loop other than using background tasks?

**Comment by chris-allnutt** [comment]
What you're looking for is gather: https://docs.python.org/3/library/asyncio-task.html#running-tasks-concurrently

If you care about the contents you can do something like the following
```python
import asyncio

await_request_1 = client.get('https://www.example.org/1')
await_request_2 = client.get('https://www.example.org/2')

await asyncio.gather(await_request_1, await_request_2)

print(await_request_1.result())
print(await_request_2.result())
```
It will allow all requests to execute concurrently

**Comment by maitham** [comment]
> What you're looking for is gather: https://docs.python.org/3/library/asyncio-task.html#running-tasks-concurrently
> 
> If you care about the contents you can do something like the following
> 
> ```python
> import asyncio
> 
> await_request_1 = client.get('https://www.example.org/1')
> await_request_2 = client.get('https://www.example.org/2')
> 
> await asyncio.gather(await_request_1, await_request_2)
> 
> print(await_request_1.result())
> print(await_request_2.result())
> ```
> 
> It will allow all requests to execute concurrently

Thanks! I probably wasn't clear enough in the description of the issue, it's less about the actually querying but more whats the best way to do it within the fastapi framework. So taking the example above is it best to define a route with `async def ...`. and bubble down all the async task. Or is ok just to run it on a separate event loop within a non-async route. I imagine the former is the best approach, but are there any downsides to doing the latter?

**Comment by tiangolo** [comment]
For the best performance, use `async`, but if you use a library that blocks, or are not completely sure, then use normal `def`.

The downside of not using `async def` is that FastAPI will run those functions in a threadpool, so you won't squeeze the maximum possible performance that `async` functions can give. It will still probably be faster than many other options. If you are having less than say 10,000 requests per second, then you are probably fine with normal `def` and not having to deal with all the details of when and when not.

Also, if you use `async def` but have some code that blocks (that is not made for `async` and `await`) that will block the event loop and degrade performance a lot. If you are not sure, go with `def`.

To learn more, check the docs: https://fastapi.tiangolo.com/async/

**Comment by maitham** [comment]
@tiangolo makes sense, in the end the functions writes to the db which is using a sqlalchemy which is non async. Ended up going the synchronous route and using requests toolbelt with a threaded map to query the external api .
